core/schema/schema.go

```diff
@@ -58,224 +57,125 @@ func New(params InitializeArgs) (*MergedSchemas, error) {
+	if err != nil {
+		return nil, err
 	}
-	return ms, nil
-}
+	api.core = &CoreMod{compiledSchema: coreSchema, introspectionJSON: coreIntrospectionJSON}
 
-func (s *MergedSchemas) getModuleSchemaView(mod *core.Module) (*schemaView, error) {
-	modDgst, err := mod.BaseDigest()
+	// the main client caller starts out the core API loaded
+	dag, err := newModDeps(ctx, api, []Mod{api.core})
 	if err != nil {
-		return nil, fmt.Errorf("failed to compute schema view digest: %w", err)
+		return nil, err
 	}
-	return s.getSchemaView(modDgst)
+	api.clientCallContext[""] = &clientCallContext{
+		dag: dag,
+	}
+
+	return api, nil
 }
 
-func (s *MergedSchemas) registerModuleFunctionCall(mod *core.Module, fnCall *core.FunctionCall) (*schemaView, digest.Digest, error) {
-	schemaView, err := s.getModuleSchemaView(mod)
-	if err != nil {
-		return nil, "", err
-	}
+type APIServer struct {
+	bk           *buildkit.Client
+	platform     specs.Platform
+	progSockPath string
+	auth         *auth.RegistryAuthProvider
+	secrets      *core.SecretStore
+	ociStore     content.Store
+	leaseManager *leaseutil.Manager
+	host         *core.Host
+	services     *core.Services
 
-	fnCallDgst, err := fnCall.Digest()
-	if err != nil {
-		return nil, "", err
-	}
+	buildCache  *core.CacheMap[uint64, *core.Container]
+	importCache *core.CacheMap[uint64, *specs.Descriptor]
 
-	dgst := digest.FromString(schemaView.viewDigest.String() + "." + fnCallDgst.String())
-	s.mu.Lock()
-	defer s.mu.Unlock()
-	s.moduleContexts[dgst] = &moduleContext{
-		module:     mod,
-		fnCall:     fnCall,
-		schemaView: schemaView,
-	}
+	// the http endpoints being served (as a map since APIs like shellEndpoint can add more)
+	endpoints  map[string]http.Handler
+	endpointMu sync.RWMutex
 
-	return schemaView, dgst, nil
-}
+	// the core API simulated as a module (intrisic dep of every user module)
+	core *CoreMod
 
-func (s *MergedSchemas) currentSchemaView(ctx context.Context) (*schemaView, error) {
-	clientMetadata, err := engine.ClientMetadataFromContext(ctx)
-	if err != nil {
-		return nil, err
-	}
+	// cache used to de-dupe loading modules from metadata
+	loadModCache *core.CacheMap[digest.Digest, *UserMod]
 
-	if clientMetadata.ModuleContextDigest == "" {
-		return s.getSchemaView("")
-	}
+	// The metadata of client calls.
+	// For the special case of the main client caller, the key is just empty string
+	// This is never explicitly deleted from; instead it will just be garbage collected
+	// when this server for the session shuts down
+	clientCallContext map[digest.Digest]*clientCallContext
+	clientCallMu      sync.RWMutex
+}
 
-	s.mu.RLock()
-	defer s.mu.RUnlock()
-	moduleContext, ok := s.moduleContexts[clientMetadata.ModuleContextDigest]
-	if !ok {
-		return nil, fmt.Errorf("module context not found")
-	}
-	return moduleContext.schemaView, nil
+type clientCallContext struct {
+	// the DAG of modules being served to this client
+	dag *ModDeps
```

Comment:
I think I agree with `deps` but could be swayed into `dag` if there is a juicy grandiose meaning to it. :stuck_out_tongue: (i.e. this is really the "one dag")

Reply:
Renamed to `deps`.

There _kind_ of is grandiose meaning; the object called `ModDeps` used to be called `ModDag` and was supposed to be the generalized DAG object (whereas `Mod` is just the vertex), but I changed it because I realized both `ModDeps` and `Mod` can really be used as a DAG; they are mostly just variations on whether there's one root or multiple.

So it's not the one DAG to rule them all, all `Mod` and `ModDeps` are manifestations of various subdags of the one- DAG to rule them all :-)

But none of that makes the name clear, so just went with the suggestion lol